### PR TITLE
GULAppDelegateSwizzler - remote notifications methods and tvOS support

### PR DESCRIPTION
**Pull request by maksymmalyhin**
Changes required for #2591.

- GULAppDelegateSwizzler - remote notifications methods support
- GULAppDelegateSwizzler tvOS support

This is a part of PR #2683. I have to merge the changes first to make sure macOS support is available for FirebaseMessaging

### CODE REVIEW

**Comment by paulb777** [comment]
Consider targeting the PR to  `mm/messaging-swizzling` instead of master to make it easier to review.

**Comment by maksymmalyhin** [comment]
@paulb777 This PR is just a part of `mm/messaging-swizzling` that has to be merged earlier to allow Cocoapods lint to pass. From that prospective, the diff between this branch and `mm/messaging-swizzling` does not make sense.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 1]
```diff
@@ -47,14 +47,39 @@ typedef BOOL (*GULRealContinueUserActivityIMP)(
...
+    id, SEL, UIApplication *, NSDictionary *, void (^)(UIBackgroundFetchResult));
+
 typedef void (^GULAppDelegateInterceptorCallback)(id<UIApplicationDelegate>);
 
 // The strings below are the keys for associated objects.
 static char const *const kGULContinueUserActivityIMPKey = "GUL_continueUserActivityIMP";
 static char const *const kGULHandleBackgroundSessionIMPKey = "GUL_handleBackgroundSessionIMP";
 static char const *const kGULOpenURLOptionsIMPKey = "GUL_openURLOptionsIMP";
+
+#if TARGET_OS_IOS
```
Comment (tejasd) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Nit: Move this to the end of this list and add a comment mentioning this method is only available on iOS and not on tvOS.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 2]
```diff
@@ -333,6 +358,7 @@ + (void)createSubclassWithObject:(id<UIApplicationDelegate>)anObject {
                                                    fromClass:realClass];
   NSValue *continueUserActivityIMPPointer = [NSValue valueWithPointer:continueUserActivityIMP];
 
+#if TARGET_OS_IOS
```
Comment (tejasd) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Nit: Move this to the end of the method.

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 3]
```diff
@@ -359,6 +386,68 @@ + (void)createSubclassWithObject:(id<UIApplicationDelegate>)anObject {
...
+                                                       fromClass:realClass];
+  NSValue *didReceiveRemoteNotificationIMPPointer =
+      [NSValue valueWithPointer:didReceiveRemoteNotificationIMP];
+
+  // For application:didReceiveRemoteNotification:fetchCompletionHandler:
+  NSValue *didReceiveRemoteNotificationWithCompletionIMPPointer;
+  SEL didReceiveRemoteNotificationWithCompletionSEL =
+      @selector(application:didReceiveRemoteNotification:fetchCompletionHandler:);
+  if ([anObject respondsToSelector:didReceiveRemoteNotificationWithCompletionSEL]) {
+    // Only add the application:didReceiveRemoteNotification:fetchCompletionHandler: method if
```
Comment (tejasd) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Nice catch!

File: GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m [hunk 4]
```diff
@@ -374,8 +463,25 @@ + (void)createSubclassWithObject:(id<UIApplicationDelegate>)anObject {
     objc_setAssociatedObject(anObject, &kGULOpenURLOptionsIMPKey, openURLOptionsIMPPointer,
                              OBJC_ASSOCIATION_RETAIN_NONATOMIC);
   }
+
+#if TARGET_OS_IOS
```
Comment (tejasd) [inline GoogleUtilities/AppDelegateSwizzler/GULAppDelegateSwizzler.m]: Nit: move to end of the list.

File: GoogleUtilities/AppDelegateSwizzler/Private/GULAppDelegateSwizzler.h [hunk 5]
```diff
@@ -55,6 +56,12 @@ typedef NSString *const GULAppDelegateInterceptorID;
  */
 + (BOOL)isAppDelegateProxyEnabled;
 
+/** Returns the current sharedApplication.
+ *
+ *  @return the current UIApplication if in an app, or nil if in extension or if it doesn't exist.
+ */
++ (nullable UIApplication *)sharedApplication;
```
Comment (tejasd) [inline GoogleUtilities/AppDelegateSwizzler/Private/GULAppDelegateSwizzler.h]: Why have you moved this to the public header? Do consumers of the AppDelegateSwizzler need this?

Comment (maksymmalyhin) [reply GoogleUtilities/AppDelegateSwizzler/Private/GULAppDelegateSwizzler.h]: @tejasd I think, I needed it in one of the versions of Messaging, but forgot to remove.

**Pull request merged**